### ISSUE TITLE
SSL accept hang with badly behaved connections

### ISSUE THREAD

**Issue by djones6** [IBM-Swift/BlueSSLService#40]
Detail in this Kitura issue: https://github.com/IBM-Swift/Kitura/issues/1143

An incoming connection that establishes a TCP session but then sends no data will block the server from accepting connections indefinitely.

Steps to recreate:
- Start the server, make some SSL requests (demonstrate it is functional)
- Connect a rogue client, eg: `telnet localhost 8443` and leave it doing nothing
- Try to make further requests. The server appears to be unresponsive

The problem seems to be that `SSL_accept` blocks, expecting to read the start of an SSL handshake from the client, which will never time out or allow the server to continue accepting connections.  In the Kitura issue above, we block here:
https://github.com/IBM-Swift/BlueSSLService/blob/335495c857b7062ecf7e228d8e021d0f7d51c5c9/Sources/SSLService.swift#L493

**Comment by mman** [comment]
SSL handshake can be potentially an expensive operation with default timeouts in tenths of seconds so doing ``SSL_accept`` should never block the main Kitura loop from accepting new connections. Otherwise the server will be still susceptible to simple DoS attacks.

Traditional ``accept``, ``fork``, ``read`` semantics with limited number of worker threads or ``fork``-ed processes should do the trick.

**Comment by billabt** [comment]
This issue is more related to the use of BlueSSLService than to BlueSSLService itself.  BlueSSLService and BlueSocket are _lower level_ functions.  The comment by @mman hits the nail on the head.  Adding a threading model at this level introduces, IMHO, potential side effects.  That should be handled at the application level.

**Comment by mman** [comment]
@billabt Just to clarify: what exactly do you mean by *should be handled at the application level*? What is the *application* here? Kitura module? Or the application that is using Kitura module inside? or Kitura-Net module?

**Comment by djones6** [comment]
To paraphrase my comments in https://github.com/IBM-Swift/Kitura-net/pull/218, the issue here is that the basic TCP accept done within `Socket.acceptClientConnection()` and the `delegate.onAccept()` that it may call are two different levels, with the latter possibly blocking.

In our case, we have a single thread that repeatedly calls `Socket.acceptClientConnection()`, which in turn calls `self.delegate?.onAccept(socket)`, and in the SSL case, calls `SSL_accept(sslConnect)` which can block.

What seems to be needed here is a decoupling of the 'Is there a connection to accept?' step from the protocol initiation step.  Kitura could continue with step 1 in a loop, but perform step 2 (which may block) on another thread (if it chooses).  I don't see how we can add a threading model around the current design.

**Comment by billabt** [comment]
The typical usage is as follows:
```
	try listener.listen(on: Int(port), maxBacklogSize: 10)

	print("Listening on port: \(port)")

	socket = try listener.acceptClientConnection()

	print("Accepted connection from: \(socket.remoteHostname) on port \(socket.remotePort), Secure? \(socket.signature!.isSecure)")
```
After the call to _listen_ returns, indicating a new connection is pending, you can start a new thread to accept the new connection…  Then return to listening again.  This will avoid tying up the main listening thread and the resultant application hang should the SSL negotiations fail or take too long…

**Comment by djones6** [comment]
@billabt It was my understanding that `listen` returns immediately, having put the socket into listening mode (ie. a server socket),  and that calling `accept` was the way to block until a connection was received:  https://www.gnu.org/software/libc/manual/html_node/Listening.html

**Comment by billabt** [comment]
@djones6: You're right...  The proper sequence would be as follows:
- **Step 1:** call _listen_ on the socket.
- **Step 2:** now call _listenerSocket.isReadableOrWritable(waitForever: true)_ to wait for a connect attempt.  **Note:** This call also supports a second parameter, _timeout_, expressed in _msec._  See the API doc for the function for more info.
- **Step 3:** when the call returns, launch a new thread that will do the _acceptClientConnection_.
- **Step 4:** then go back and repeat **Steps 2 & 3** to wait for additional connections.

**Comment by djones6** [comment]
@billabt This doesn't seem to work either.  `isReadableOrWritable(waitForever: true)` throws `SOCKET_ERR_NOT_CONNECTED` when I call it on the `listenerSocket`.

**Comment by billabt** [comment]
@djones6:  I'll have to make some changes to get this to work this way. Reopening this issue.

**Comment by djones6** [comment]
@billabt I have a working solution for this in IBM-Swift/BlueSocket#85 and IBM-Swift/Kitura-net#219

**Comment by billabt** [comment]
@djones6:  I'll take a look at your PR. I'd like to avoid adding the new API if possible. What happens if the app never invokes the new API? The delegate will never be called.

**Comment by djones6** [comment]
That's true, however in order to avoid invoking the delegate the app must supply the optional parameter `invokeDelegate: false`.  That puts the onus on the application to ensure it does call the delegate, which IMHO is perfectly reasonable.

**Comment by billabt** [comment]
@djones6: True, that's a reasonable assumption.  If we accept this PR, I'll have to bump the **MINOR** version number since the API changed.  Does that present a problem for Kitura?

**Comment by billabt** [comment]
@djones6:  I'm going to accept the PR as soon as CI finishes but I've got to make some changes.  As it stands right now, the _invokeDelegateOnAccept_ method doesn't protect against being called multiple times or with the wrong socket.  I've also got to update the README to take into account the new parameter and the new API.

**Comment by billabt** [comment]
Fixed in v0.12.66 of _BlueSocket_.
